Core/clim-basic/input.lisp

```diff
@@ -409,22 +408,29 @@ use condition-variables nor locks."))
   (check-schedule queue)
   (not (null (event-queue-head queue))))
 
-(defmethod event-queue-listen-or-wait ((queue concurrent-event-queue) &key timeout)
+(defmethod event-queue-listen-or-wait ((queue concurrent-event-queue)
+                                       &key timeout wait-function)
   (do-port-force-output queue)
   ;; We need to LOOP because of possible spurious wakeup (sbcl/bt quirk).
   (loop
+     with pred = (or wait-function #'(lambda () (event-queue-head queue)))
      with lock = (event-queue-lock queue)
      with cv = (event-queue-processes queue)
      with timeout-time = (and timeout (+ timeout (now)))
      as decay = (compute-decay timeout-time (event-schedule-time queue))
      do
-       (when decay (maxf decay 0))
+     ;; We CLAMPF decay when wait-function is present to ensure that
+     ;; we don't get stuck until next event arrives (or the timeout
+     ;; happens). It is busy wait with a lousy grain. -- jd 2019-06-06
+       (when decay (if wait-function
+                       (clampf decay 0 0.01)
+                       (maxf decay 0)))
```

Comment:
compute-decay computes the time between two timestamps which we need to wait (and returns negative value when we are past due). It makes sense to not get rid of that information. otoh timeout can't be specified to have negative value (hence clampf/maxf)

Reply:
Does this actually happen? If the decay is 0, then we've already passed the time where the waiting should end, and as such it wouldn't be called again.

Reply:
Since the function is always used to compute the wait time (and as such, always clamped to 0), perhaps there should be a function whose purpose it is to do exactly this. However, we're getting into hair splitting territory here, and it's not very important really.